rfd/0105-headless-authentication.md

```diff
@@ -0,0 +1,406 @@
+    Username string `json:"username"`
+    // Cert is a PEM encoded SSH certificate signed by SSH certificate authority
+    Cert []byte `json:"cert"`
+    // TLSCertPEM is a PEM encoded TLS certificate signed by TLS certificate authority
+    TLSCert []byte `json:"tls_cert"`
+    // HostSigners is a list of signing host public keys trusted by proxy
+    HostSigners []TrustedCerts `json:"host_signers"`
+}
+```
+
+#### `rpc CreateHeadlessAuthentication`
+
+This endpoint is used by the Teleport Proxy to create a new headless auth request.
+
+It is authorized for roles with `create` permissions on the `headless_authentication` resource kind, which is only given to the Proxy role by default.
+
+#### `rpc GetHeadlessAuthentication`
+
+This endpoint is used by Teleport clients to retrieve headless login request details before prompting the user for approval/denial.
+
+It is authorized for roles with `read` permissions on the `headless_authentication` resource kind, or for the user who requested headless authentication.
+
+#### `rpc UpdateHeadlessAuthenticationState`
+
+This endpoint is used by Teleport clients to update headless login request state to approved or denied. 
+
+It is only authorized authorized for the user who requested headless authentication. Additionally, when updating the state to `APPROVED`, the client must provide a valid MFA challenge response for the user. An MFA challenge can be requested from the existing rpc `CreateAuthenticateChallenge`.
+
+#### `rpc GenerateUserCerts`
+
+This is an existing endpoint used to generate user certs. We will add `headless_authentication_id` field to retrieve request details from the headless auth request saved on the Auth server. We will add the `headless_authentication_token` field to authorize the certificate generation for the headless authentication.
+
+```proto
+// UserCertRequest specifies certificate-generation parameters
+// for a user.
+message UserCertsRequest {
+  ...
+  // HeadlessAuthenticationID is the id of a headless authentication entity.
+  // When set, the actual cert request will be formed from the headless authentication
+  // details saved on the Auth server.
+  string headless_authentication_id = 17;
+
+  // HeadlessAuthenticationToken is a secret token used to authorize generating
+  // certificates for a successful headless authentication. This is only known by 
+  // the Proxy Server performing the headless authentication flow and the Auth server.
+  string headless_authentication_token = 18;
+}
+```
+
+### UX
+
+#### `tsh --headless`
+
+We will add a new `--headless` flag to `tsh` which can be used to authenticate for a single `tsh` request. When this flag is provided, `tsh` will prompt the user to complete headless authentication on their local machine from the URL `https://proxy.example.com/headless/<id>/approve`. Once the user completes local authentication, `tsh` will receive credentials to complete the request.
```

Comment:
I'd mention `TELEPORT_HEADLESS` here too.

Reply:
This is mentioned in the section below. I can move that section into this one if you think that's better.